File: src/main/java/net/softbell/bsh/iot/service/v1/IotTriggerServiceV1.java
package net.softbell.bsh.iot.service.v1;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Optional;

import javax.transaction.Transactional;

import org.springframework.security.core.Authentication;
import org.springframework.stereotype.Service;

import lombok.AllArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import net.softbell.bsh.domain.EnableStatusRule;
import net.softbell.bsh.domain.TriggerLastStatusRule;
import net.softbell.bsh.domain.TriggerStatusRule;
import net.softbell.bsh.domain.entity.Member;
import net.softbell.bsh.domain.entity.NodeAction;
import net.softbell.bsh.domain.entity.NodeItem;
import net.softbell.bsh.domain.entity.NodeTrigger;
import net.softbell.bsh.domain.entity.NodeTriggerAction;
import net.softbell.bsh.domain.repository.NodeActionRepo;
import net.softbell.bsh.domain.repository.NodeItemRepo;
import net.softbell.bsh.domain.repository.NodeTriggerActionRepo;
import net.softbell.bsh.domain.repository.NodeTriggerRepo;
import net.softbell.bsh.dto.request.IotTriggerActionDto;
import net.softbell.bsh.dto.request.IotTriggerDto;
import net.softbell.bsh.iot.component.v1.IotAuthCompV1;
import net.softbell.bsh.iot.component.v1.IotChannelCompV1;
import net.softbell.bsh.iot.component.v1.IotTriggerParserCompV1;
import net.softbell.bsh.service.MemberService;

/**
 * @Author : Bell(bell@softbell.net)
 * @Description : IoT Node Trigger 서비스
 */
@Slf4j
@AllArgsConstructor
@Service
public class IotTriggerServiceV1
{
	// Global Field
	private final MemberService memberService;
	private final IotActionServiceV1 iotActionService;
	
	private final IotTriggerParserCompV1 iotTriggerParserComp;
	private final IotChannelCompV1 iotChannelCompV1;
	private final IotAuthCompV1 iotAuthCompV1;
	
	private final NodeTriggerRepo nodeTriggerRepo;
	private final NodeTriggerActionRepo nodeTriggerActionRepo;
	private final NodeActionRepo nodeActionRepo;
	private final NodeItemRepo nodeItemRepo;
	
	public List<NodeTrigger> getAllTriggers(Authentication auth)
	{
		return nodeTriggerRepo.findAll();
	}
	
	public NodeTrigger getTrigger(Authentication auth, long triggerId)
	{
		// Field
		Optional<NodeTrigger> optNodeTrigger;
		
		// Init
		optNodeTrigger = nodeTriggerRepo.findById(triggerId);
		
		// Exception
		if (!optNodeTrigger.isPresent())
			return null;
		
		// Return
		return optNodeTrigger.get();
	}
	
	@Transactional
	public boolean createTrigger(Authentication auth, IotTriggerDto iotTriggerDto)
	{
		// Exception
		if (iotTriggerDto == null)
			return false;
		
		// Field
		Member member;
		NodeTrigger nodeTrigger;
		HashMap<Long, IotTriggerActionDto> mapAction;
		List<NodeTriggerAction> listAction;
		EnableStatusRule enableStatus;
		
		// Init
		listAction = new ArrayList<NodeTriggerAction>();
		mapAction = iotTriggerDto.getMapAction();
		member = memberService.getMember(auth.getName());
		
		if (iotTriggerDto.isEnableStatus())
			enableStatus = EnableStatusRule.ENABLE;
		else
			enableStatus = EnableStatusRule.DISABLE;
		
		// Exception
		if (member == null)
			return false;
		
		// Data Process - Node Trigger
		nodeTrigger = NodeTrigger.builder()
										.enableStatus(enableStatus)
										.description(iotTriggerDto.getDescription())
										.expression(iotTriggerDto.getExpression())
										.member(member)
										.lastStatus(TriggerLastStatusRule.WAIT)
										.build();
		
		// DB - Node Trigger Save
		nodeTriggerRepo.save(nodeTrigger);
		
		
		// Data Process - Node Trigger Action
		if (mapAction != null)
		{
			mapAction.forEach((key, value) ->
			{
				// Field
				Optional<NodeAction> optNodeAction;
				NodeTriggerAction nodeTriggerAction;
				
				// Init
				optNodeAction = nodeActionRepo.findById(key);
				
				// Build
				if (optNodeAction.isPresent())
				{
					nodeTriggerAction = NodeTriggerAction.builder()
															.nodeTrigger(nodeTrigger)
															.nodeAction(optNodeAction.get())
																.build();
					if (value.isEventError() && value.isEventOccur() && value.isEventRestore())
						nodeTriggerAction.setTriggerStatus(TriggerStatusRule.ALL);
					else if (value.isEventOccur() && value.isEventRestore())
						nodeTriggerAction.setTriggerStatus(TriggerStatusRule.OCCUR_AND_RESTORE);
					else if (value.isEventError())
						nodeTriggerAction.setTriggerStatus(TriggerStatusRule.ERROR);
					else if (value.isEventOccur())
						nodeTriggerAction.setTriggerStatus(TriggerStatusRule.OCCUR);
					else if (value.isEventRestore())
						nodeTriggerAction.setTriggerStatus(TriggerStatusRule.RESTORE);
					
					// List Add
					listAction.add(nodeTriggerAction);
				}
			});
			nodeTrigger.setNodeTriggerActions(listAction);
		}
		
		// DB - Node Trigger Action Save
		nodeTriggerActionRepo.saveAll(listAction);
		
		
		// Return
		return true;
	}
	
	@Transactional
	public boolean modifyTrigger(Authentication auth, long triggerId, IotTriggerDto iotTriggerDto)
	{
		// Exception
		if (iotTriggerDto == null)
			return false;
		
		// Field
//		Member member;
		Optional<NodeTrigger> optNodeTrigger;
		NodeTrigger nodeTrigger;
		HashMap<Long, IotTriggerActionDto> mapAction;
		List<NodeTriggerAction> listAction;
		EnableStatusRule enableStatus;
		
		// Init
		listAction = new ArrayList<NodeTriggerAction>();
		mapAction = iotTriggerDto.getMapAction();
//		member = memberService.getMember(auth.getName());
		optNodeTrigger = nodeTriggerRepo.findById(triggerId);
		
		if (iotTriggerDto.isEnableStatus())
			enableStatus = EnableStatusRule.ENABLE;
		else
			enableStatus = EnableStatusRule.DISABLE;
		
		// Exception
		if (/*member == null || */!optNodeTrigger.isPresent())
			return false;
		
		// Init
		nodeTrigger = optNodeTrigger.get();
		
		// DB - Trigger Action Clear
		nodeTriggerActionRepo.deleteAll(nodeTrigger.getNodeTriggerActions());
		
		// Data Process - Node Trigger
		nodeTrigger.setEnableStatus(enableStatus);
		nodeTrigger.setDescription(iotTriggerDto.getDescription());
		nodeTrigger.setExpression(iotTriggerDto.getExpression());
		
		// DB - Node Trigger Save
		nodeTriggerRepo.save(nodeTrigger);
		
		
		// Data Process - Node Trigger Action
		if (mapAction != null)
		{
			mapAction.forEach((key, value) ->
			{
				// Field
				Optional<NodeAction> optNodeAction;
				NodeTriggerAction nodeTriggerAction;
				
				// Init
				optNodeAction = nodeActionRepo.findById(key);
				
				// Build
				if (optNodeAction.isPresent())
				{
					nodeTriggerAction = NodeTriggerAction.builder()
															.nodeTrigger(nodeTrigger)
															.nodeAction(optNodeAction.get())
																.build();
					if (value.isEventError() && value.isEventOccur() && value.isEventRestore())
						nodeTriggerAction.setTriggerStatus(TriggerStatusRule.ALL);
					else if (value.isEventOccur() && value.isEventRestore())
						nodeTriggerAction.setTriggerStatus(TriggerStatusRule.OCCUR_AND_RESTORE);
					else if (value.isEventError())
						nodeTriggerAction.setTriggerStatus(TriggerStatusRule.ERROR);
					else if (value.isEventOccur())
						nodeTriggerAction.setTriggerStatus(TriggerStatusRule.OCCUR);
					else if (value.isEventRestore())
						nodeTriggerAction.setTriggerStatus(TriggerStatusRule.RESTORE);
					
					// List Add
					listAction.add(nodeTriggerAction);
				}
			});
			nodeTrigger.setNodeTriggerActions(listAction);
		}
		
		// DB - Node Trigger Action Save
		nodeTriggerActionRepo.saveAll(listAction);
		
		
		// Return
		return true;
	}
	
	@Transactional
	public boolean setTriggerEnableStatus(Authentication auth, long triggerId, boolean status)
	{
		// Field
		Optional<NodeTrigger> optNodeTrigger;
		NodeTrigger nodeTrigger;
		
		// Init
		optNodeTrigger = nodeTriggerRepo.findById(triggerId);
		
		// Exception
		if (!optNodeTrigger.isPresent())
			return false;
		
		// Load
		nodeTrigger = optNodeTrigger.get();
		
		// DB - Update
		if (status)
			nodeTrigger.setEnableStatus(EnableStatusRule.ENABLE);
		else
			nodeTrigger.setEnableStatus(EnableStatusRule.DISABLE);
		
		// Return
		return true;
	}
	
	@Transactional
	public boolean deleteTrigger(Authentication auth, long triggerId)
	{
		// Field
		Optional<NodeTrigger> optNodeTrigger;
		NodeTrigger nodeTrigger;
		
		// Init
		optNodeTrigger = nodeTriggerRepo.findById(triggerId);
		
		// Exception
		if (!optNodeTrigger.isPresent())
			return false;
		nodeTrigger = optNodeTrigger.get();
		
		// DB - Delete
		for (NodeTriggerAction entity : nodeTrigger.getNodeTriggerActions())
			nodeTriggerActionRepo.delete(entity);
		nodeTriggerRepo.delete(nodeTrigger);
		
		// Return
		return true;
	}
	
	@Transactional
	public boolean procTrigger(NodeItem nodeItem)
	{
		// Field
		boolean isSuccess;
		List<NodeTrigger> listNodeTrigger;
		List<NodeAction> listNodeAction; 
		
		// Init
		isSuccess = true;
		listNodeTrigger = iotTriggerParserComp.convTrigger(nodeItem);
		listNodeAction = new ArrayList<NodeAction>();
		
		// Process
		for (NodeTrigger nodeTrigger : listNodeTrigger)
		{
			// Field
			Boolean isOccur;
			
			// Init
			isOccur = iotTriggerParserComp.parseEntity(nodeTrigger);
			
			// Exception
			if ((nodeTrigger.getLastStatus() == TriggerLastStatusRule.ERROR && isOccur == null) ||
					(nodeTrigger.getLastStatus() == TriggerLastStatusRule.OCCUR && isOccur) ||
					(nodeTrigger.getLastStatus() == TriggerLastStatusRule.RESTORE && !isOccur))
				continue;
			
			// DB - Update
			if (isOccur == null)
				nodeTrigger.setLastStatus(TriggerLastStatusRule.ERROR); // DB Update - Error
			else if (isOccur)
				nodeTrigger.setLastStatus(TriggerLastStatusRule.OCCUR); // Occur
			else
				nodeTrigger.setLastStatus(TriggerLastStatusRule.RESTORE); // Restore
			
			// Load
			listNodeAction.addAll(iotTriggerParserComp.getTriggerAction(nodeTrigger));
		}
		
		// Exec Action
		for (NodeAction nodeAction : listNodeAction)
			if (!iotActionService.execAction(nodeAction))
				isSuccess = false;
		
		// Return
		return isSuccess;
	}
	
	
}

File: src/main/kotlin/net/softbell/bsh/iot/service/v1/IotTriggerServiceV1.kt
package net.softbell.bsh.iot.service.v1

import net.softbell.bsh.domain.EnableStatusRule
import net.softbell.bsh.domain.TriggerLastStatusRule
import net.softbell.bsh.domain.TriggerStatusRule
import net.softbell.bsh.domain.entity.NodeItem
import net.softbell.bsh.domain.entity.NodeTrigger
import net.softbell.bsh.domain.entity.NodeTriggerAction
import net.softbell.bsh.domain.repository.NodeActionRepo
import net.softbell.bsh.domain.repository.NodeTriggerActionRepo
import net.softbell.bsh.domain.repository.NodeTriggerRepo
import net.softbell.bsh.dto.request.IotTriggerActionDto
import net.softbell.bsh.dto.request.IotTriggerDto
import net.softbell.bsh.iot.component.v1.IotTriggerParserCompV1
import net.softbell.bsh.service.MemberService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.security.core.Authentication
import org.springframework.stereotype.Service
import java.util.*
import java.util.function.BiConsumer
import javax.transaction.Transactional

/**
 * @author : Bell(bell@softbell.net)
 * @description : IoT Node Trigger 서비스
 */
@Service
class IotTriggerServiceV1 {
    // Global Field
    @Autowired private lateinit var memberService: MemberService
    @Autowired private lateinit var iotActionService: IotActionServiceV1
    @Autowired private lateinit var iotTriggerParserComp: IotTriggerParserCompV1
    @Autowired private lateinit var nodeTriggerRepo: NodeTriggerRepo
    @Autowired private lateinit var nodeTriggerActionRepo: NodeTriggerActionRepo
    @Autowired private lateinit var nodeActionRepo: NodeActionRepo

    fun getAllTriggers(auth: Authentication): List<NodeTrigger> {
        // Init
        val member = memberService.getMember(auth.name) ?: return emptyList()

        // Return
        return if (memberService.isAdmin(member))
            nodeTriggerRepo.findAll()
        else
            nodeTriggerRepo.findByMember(member)
    }

    fun getTrigger(auth: Authentication, triggerId: Long): NodeTrigger? {
        // Init
        val optNodeTrigger = nodeTriggerRepo.findById(triggerId)

        // Return
        return if (optNodeTrigger.isPresent)
            optNodeTrigger.get()
        else
            null
    }

    @Transactional
    fun createTrigger(auth: Authentication, iotTriggerDto: IotTriggerDto): Boolean {
        // Init
        val listAction: MutableList<NodeTriggerAction> = ArrayList()
        val mapAction = iotTriggerDto.mapAction
        val member = memberService.getMember(auth.name) ?: return false
        val enableStatus = if (iotTriggerDto.enableStatus)
            EnableStatusRule.ENABLE
        else
            EnableStatusRule.DISABLE

        // Data Process - Node Trigger
        val nodeTrigger = NodeTrigger(
                enableStatus = enableStatus,
                description = iotTriggerDto.description,
                expression = iotTriggerDto.expression,
                member = member,
                lastStatus = TriggerLastStatusRule.WAIT
        )

        // DB - Node Trigger Save
        nodeTriggerRepo.save(nodeTrigger)


        // Data Process - Node Trigger Action
        if (mapAction != null) {
            mapAction.forEach(BiConsumer { key: Long, value: IotTriggerActionDto ->
                // Init
                val optNodeAction = nodeActionRepo.findById(key)

                // Build
                if (optNodeAction.isPresent) {
                    val nodeTriggerAction = NodeTriggerAction(
                            nodeTrigger = nodeTrigger,
                            nodeAction = optNodeAction.get(),
                            triggerStatus = TriggerStatusRule.RESTORE // TEMP
                    )

                    if (value.eventError && value.eventOccur && value.eventRestore)
                        nodeTriggerAction.triggerStatus = TriggerStatusRule.ALL
                    else if (value.eventOccur && value.eventRestore)
                        nodeTriggerAction.triggerStatus = TriggerStatusRule.OCCUR_AND_RESTORE
                    else if (value.eventError)
                        nodeTriggerAction.triggerStatus = TriggerStatusRule.ERROR
                    else if (value.eventOccur)
                        nodeTriggerAction.triggerStatus = TriggerStatusRule.OCCUR
                    else if (value.eventRestore)
                        nodeTriggerAction.triggerStatus = TriggerStatusRule.RESTORE

                    // List Add
                    listAction.add(nodeTriggerAction)
                }
            })
            nodeTrigger.nodeTriggerActions = listAction
        }

        // DB - Node Trigger Action Save
        nodeTriggerActionRepo.saveAll(listAction)


        // Return
        return true
    }

    @Transactional
    fun modifyTrigger(auth: Authentication, triggerId: Long, iotTriggerDto: IotTriggerDto): Boolean {
        // Init
        val listAction: MutableList<NodeTriggerAction> = ArrayList()
        val mapAction = iotTriggerDto.mapAction
        val optNodeTrigger = nodeTriggerRepo.findById(triggerId)
        val enableStatus = if (iotTriggerDto.enableStatus)
            EnableStatusRule.ENABLE
        else
            EnableStatusRule.DISABLE

        // Exception
        if (!optNodeTrigger.isPresent)
            return false

        // Init
        val nodeTrigger = optNodeTrigger.get()

        // DB - Trigger Action Clear
        nodeTriggerActionRepo.deleteAll(nodeTrigger.nodeTriggerActions)

        // Data Process - Node Trigger
        nodeTrigger.enableStatus = enableStatus
        nodeTrigger.description = iotTriggerDto.description
        nodeTrigger.expression = iotTriggerDto.expression

        // DB - Node Trigger Save
        nodeTriggerRepo.save(nodeTrigger)


        // Data Process - Node Trigger Action
        if (mapAction != null) {
            mapAction.forEach(BiConsumer { key: Long, value: IotTriggerActionDto ->
                // Init
                val optNodeAction = nodeActionRepo.findById(key)

                // Build
                if (optNodeAction.isPresent) {
                    val nodeTriggerAction = NodeTriggerAction(
                            nodeTrigger = nodeTrigger,
                            nodeAction = optNodeAction.get(),
                            triggerStatus = TriggerStatusRule.RESTORE // TEMP
                    )

                    if (value.eventError && value.eventOccur && value.eventRestore)
                        nodeTriggerAction.triggerStatus = TriggerStatusRule.ALL
                    else if (value.eventOccur && value.eventRestore)
                        nodeTriggerAction.triggerStatus = TriggerStatusRule.OCCUR_AND_RESTORE
                    else if (value.eventError)
                        nodeTriggerAction.triggerStatus = TriggerStatusRule.ERROR
                    else if (value.eventOccur)
                        nodeTriggerAction.triggerStatus = TriggerStatusRule.OCCUR
                    else if (value.eventRestore)
                        nodeTriggerAction.triggerStatus = TriggerStatusRule.RESTORE

                    // List Add
                    listAction.add(nodeTriggerAction)
                }
            })
            nodeTrigger.nodeTriggerActions = listAction
        }

        // DB - Node Trigger Action Save
        nodeTriggerActionRepo.saveAll(listAction)

        // Return
        return true
    }

    @Transactional
    fun setTriggerEnableStatus(auth: Authentication, triggerId: Long, status: Boolean): Boolean {
        // Init
        val optNodeTrigger = nodeTriggerRepo.findById(triggerId)

        // Exception
        if (!optNodeTrigger.isPresent)
            return false

        // Load
        val nodeTrigger = optNodeTrigger.get()

        // DB - Update
        if (status)
            nodeTrigger.enableStatus = EnableStatusRule.ENABLE
        else
            nodeTrigger.enableStatus = EnableStatusRule.DISABLE

        // Return
        return true
    }

    @Transactional
    fun deleteTrigger(auth: Authentication, triggerId: Long): Boolean {
        // Init
        val optNodeTrigger = nodeTriggerRepo.findById(triggerId)

        // Exception
        if (!optNodeTrigger.isPresent)
            return false
        val nodeTrigger = optNodeTrigger.get()

        // DB - Delete
        for (entity in nodeTrigger.nodeTriggerActions)
            nodeTriggerActionRepo.delete(entity)
        nodeTriggerRepo.delete(nodeTrigger)

        // Return
        return true
    }

    @Transactional
    fun procTrigger(nodeItem: NodeItem): Boolean {
        // Init
        var isSuccess = true
        val listNodeTrigger = iotTriggerParserComp.convTrigger(nodeItem)

        // Process
        for (nodeTrigger in listNodeTrigger) {
            // Init
            val isOccur = iotTriggerParserComp.parseEntity(nodeTrigger)

            // Exception
            if (nodeTrigger.lastStatus == TriggerLastStatusRule.UNKNOWN && isOccur == null ||
                    nodeTrigger.lastStatus == TriggerLastStatusRule.OCCUR && isOccur!! ||
                    nodeTrigger.lastStatus == TriggerLastStatusRule.RESTORE && !isOccur!!)
                continue

            // DB - Update
            when {
                isOccur == null -> nodeTrigger.lastStatus = TriggerLastStatusRule.UNKNOWN // DB Update - Error
                isOccur -> nodeTrigger.lastStatus = TriggerLastStatusRule.OCCUR // Occur
                else -> nodeTrigger.lastStatus = TriggerLastStatusRule.RESTORE // Restore
            }

            // Exec Action
            for (nodeAction in iotTriggerParserComp.getTriggerAction(nodeTrigger))
                if (!iotActionService.execAction(nodeAction, nodeTrigger.member))
                    isSuccess = false
        }

        // Return
        return isSuccess
    }
}